gtk4/src/subclass/entry_buffer.rs

```diff
@@ -0,0 +1,242 @@
+            let data = T::type_data();
+            let parent_class = data.as_ref().get_parent_class() as *mut ffi::GtkEntryBufferClass;
+            let f = (*parent_class)
+                .insert_text
+                .expect("No parent class impl for \"insert_text\"");
+
+            f(
+                entry_buffer
+                    .unsafe_cast_ref::<EntryBuffer>()
+                    .to_glib_none()
+                    .0,
+                position,
+                chars.to_glib_none().0,
+                chars.len() as u32,
+            )
+        }
+    }
+
+    fn parent_inserted_text(&self, entry_buffer: &Self::Type, position: u32, chars: &str) {
+        unsafe {
+            let data = T::type_data();
+            let parent_class = data.as_ref().get_parent_class() as *mut ffi::GtkEntryBufferClass;
+            if let Some(f) = (*parent_class).inserted_text {
+                f(
+                    entry_buffer
+                        .unsafe_cast_ref::<EntryBuffer>()
+                        .to_glib_none()
+                        .0,
+                    position,
+                    chars.to_glib_none().0,
+                    chars.len() as u32,
+                )
+            }
+        }
+    }
+}
+
+unsafe impl<T: EntryBufferImpl> IsSubclassable<T> for EntryBuffer {
+    fn override_vfuncs(class: &mut glib::Class<Self>) {
+        <Object as IsSubclassable<T>>::override_vfuncs(class);
+
+        let klass = class.as_mut();
+        klass.delete_text = Some(entry_buffer_delete_text::<T>);
+        klass.deleted_text = Some(entry_buffer_deleted_text::<T>);
+        klass.get_length = Some(entry_buffer_get_length::<T>);
+        klass.get_text = Some(entry_buffer_get_text::<T>);
+        klass.insert_text = Some(entry_buffer_insert_text::<T>);
+        klass.inserted_text = Some(entry_buffer_inserted_text::<T>);
+    }
+}
+
+unsafe extern "C" fn entry_buffer_delete_text<T: EntryBufferImpl>(
+    ptr: *mut ffi::GtkEntryBuffer,
+    position: u32,
+    n_chars: u32,
+) -> u32 {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<EntryBuffer> = from_glib_borrow(ptr);
+
+    imp.delete_text(wrap.unsafe_cast_ref(), position, n_chars)
+}
+
+unsafe extern "C" fn entry_buffer_deleted_text<T: EntryBufferImpl>(
+    ptr: *mut ffi::GtkEntryBuffer,
+    position: u32,
+    n_chars: u32,
+) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<EntryBuffer> = from_glib_borrow(ptr);
+
+    imp.deleted_text(wrap.unsafe_cast_ref(), position, n_chars)
+}
+
+unsafe extern "C" fn entry_buffer_get_text<T: EntryBufferImpl>(
+    ptr: *mut ffi::GtkEntryBuffer,
+    _n_bytes: *mut usize,
+) -> *const libc::c_char {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<EntryBuffer> = from_glib_borrow(ptr);
+
+    imp.get_text(wrap.unsafe_cast_ref()).to_glib_full()
```

Comment:
Welcome to stupid C APIs :) The C function is `transfer none` in its return value. That means that you are currently creating a new copy of the string and then leaking it, which is not optimal.

You could store the `to_glib_full()` return value in `g_object_set_qdata_full()` (use `g_free()` as free function) and replace that on every call. Then it's at least kept valid until the next call to `get_text()`, which should be sufficient here.

Reply:
I think there's an example of this pattern in `gio` in `IOStream` for the getters of the input/output stream.

Reply:
> You could store the `to_glib_full()` return value in `g_object_set_qdata_full()` (use `g_free()` as free function) and replace that on every call. Then it's at least kept valid until the next call to `get_text()`, which should be sufficient here.

We already have safe bindings for `set_qdata` =p